metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -64,21 +65,18 @@
         private void update(long value, long timestampNanos) {
             double itemWeight = weight(timestampNanos - startTick);
             double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
+            long currentCount = count.get();
+            if (currentCount < size) {
+                addSample(priority, value, itemWeight);
+            } else if (values.firstKey() < priority) {
+                addSample(priority, value, itemWeight);
+            }
+        }
 
-            long newCount = count.incrementAndGet();
-            if (newCount <= size || values.isEmpty()) {
-                values.put(priority, new WeightedSample(value, itemWeight));
-            } else {
-                Double first = values.firstKey();
-                if (first < priority
-                        // Optimization: Avoid WeightedSample allocation in the hot path when priority is lower than
-                        // the existing minimum.
-                        && values.putIfAbsent(priority, new WeightedSample(value, itemWeight)) == null) {
-                    // Always remove an item
-                    while (values.remove(first) == null) {
-                        first = values.firstKey();
-                    }
-                }
+        private void addSample(double priority, long value, double itemWeight) {
+            if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight)) == null
+                    && count.incrementAndGet() > size) {
```

Comment:
count.increment should be done earlier

Reply:
I think this fixes a bug in the original implementation in which the first `size` put operations could (though _very_ unlikely) roll the same priority for every element.

Reply:
Right, you might be correct about that!